peer/x/peerheap/list.go

```diff
@@ -0,0 +1,242 @@
+	ps, ok := pl.byIdentifier[pid.Identifier()]
+	if !ok {
+		return peer.ErrPeerRemoveNotInList(pid.Identifier())
+	}
+
+	if err := pl.byScore.validate(ps); err != nil {
+		return err
+	}
+
+	err := pl.transport.ReleasePeer(pid, ps)
+	delete(pl.byIdentifier, pid.Identifier())
+	pl.byScore.delete(ps.idx)
+	ps.list = nil
+	return err
+}
+
+func (pl *List) clearPeers() error {
+	pl.mu.Lock()
+	defer pl.mu.Unlock()
+
+	var errs []error
+
+	for {
+		ps, ok := pl.byScore.peekPeer()
+		if !ok {
+			break
+		}
+		if err := pl.releasePeer(ps.id); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return yerrors.MultiError(errs)
+}
+
+// Choose satisfies peer.Chooser, providing a single peer for a request, a
+// callback for when the request is finished, or an error if it fails.
+// The choose method takes a context that must have a deadline.
+// Choose resepects this deadline, waiting for an available peer until the
+// deadline.
+// The peer heap does not use the given *transport.Request and can safely
+// receive nil.
+func (pl *List) Choose(ctx context.Context, _ *transport.Request) (peer.Peer, func(error), error) {
+	if !pl.IsRunning() {
+		return nil, nil, peer.ErrPeerListNotStarted("PeerHeap")
+	}
+
+	for {
+		if ps, ok := pl.get(); ok {
+			pl.notifyPeerAvailable()
+			ps.peer.StartRequest()
+			return ps.peer, ps.boundFinish, nil
+		}
+
+		if err := pl.waitForPeerAvailableEvent(ctx); err != nil {
+			return nil, nil, err
+		}
+	}
+}
+
+func (pl *List) get() (*peerScore, bool) {
+	pl.mu.Lock()
+	defer pl.mu.Unlock()
+
+	ps, ok := pl.byScore.popPeer()
+	if !ok {
+		return nil, false
+	}
+
+	// Note: We push the peer back to reset the "next" counter.
+	// This gives us round-robin behavior.
+	pl.byScore.pushPeer(ps)
+
+	return ps, ps.status.ConnectionStatus == peer.Available
+}
+
+// waitForPeerAvailableEvent waits until a peer is added to the peer list or the
+// given context finishes.
+// Must NOT be run in a mutex.Lock()
+func (pl *List) waitForPeerAvailableEvent(ctx context.Context) error {
+	if _, ok := ctx.Deadline(); !ok {
+		return peer.ErrChooseContextHasNoDeadline("PeerHeap")
+	}
+
+	select {
+	case <-pl.peerAvailableEvent:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// notifyPeerAvailable writes to a channel indicating that a Peer is currently
+// available for requests
+func (pl *List) notifyPeerAvailable() {
+	select {
+	case pl.peerAvailableEvent <- struct{}{}:
+	default:
+	}
+}
+
+// NotifyStatusChanged receives notifications when a peer becomes available,
+// connected, unavailable, or when its pending request count changes.
+// This method satisfies peer.Subscriber and is only used for tests, since
+// the peer heap has a subscriber for each invividual peer.
+func (pl *List) NotifyStatusChanged(pid peer.Identifier) {
+	pl.mu.Lock()
+	ps := pl.byIdentifier[pid.Identifier()]
+	pl.mu.Unlock()
+	ps.NotifyStatusChanged(pid)
+}
+
+func (pl *List) notifyStatusChanged(ps *peerScore) {
```

Comment:
That is a little weird, I renamed it to `peerScoreChanged` to make it a little clearer.